### ISSUE TITLE
Deprecate AuthenticatorAttachment in favor of PublicKeyCredentialHints. 

### ISSUE THREAD

**Issue by akshayku** [w3c/webauthn#2053]
## Background

We have introduced PublicKeyCredentialHints for RP to better convey intention of which transport is preferred for both credential creation and authentication. Previously, AuthenticatorAttachment was used to convey similar intention. 

However, AuthenticatorAttachment has a side affect during credential creation. It excludes certain authenticators and the definition of platform vs cross-platform has been murky for some time since the introduction of hybrid transport. It leads to market fragmentation and UI differences between platforms. 

## Proposed Change
Deprecate AuthenticatorAttachment in favor of PublicKeyCredentialHints in the spec. For backwards compatibilty when only attachment is provided, we can map those values to corresponding PublicKeyCredentialHints.

**Comment by timcappalli** [comment]
We discussed this on a WG call a few weeks back, and I think there was loose agreement to do this as part of Level 4 (e.g. replacement added in L3, deprecate old way in L4).

**Comment by MasterKale** [comment]
We attempted to hedge this by establishing in L3 that hints override attachment:

> Hints MAY contradict information contained in credential [transports](https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-transports) and [authenticatorAttachment](https://w3c.github.io/webauthn/#dom-authenticatorselectioncriteria-authenticatorattachment). When this occurs, the hints take precedence. (Note that [transports](https://w3c.github.io/webauthn/#dom-publickeycredentialdescriptor-transports) values are not provided when using [discoverable credentials](https://w3c.github.io/webauthn/#discoverable-credential), leaving hints as the only avenue for expressing some aspects of such a request.)

Deprecating in L4 definitely makes sense.



> But I think we need to address RPs who still want to specify local vs remote, without having to worry about all the options. This could be addressed by just adding two additional options to hints (e.g. `local-client`, `remote-authenticator`).

@timcappalli Are you thinking about auth here? Because `client-device` and `hybrid` hints should solve this for registration...

**Comment by timcappalli** [comment]
> @timcappalli Are you thinking about auth here? Because client-device and hybrid hints should solve this for registration...

I updated my comment. We talked about adding something equivalent to `cross-device` to cover both hybrid and security-keys and any other future non-local option.

**Comment by MasterKale** [comment]
> I updated my comment. We talked about adding something equivalent to `cross-device` to cover both hybrid and security-keys and any other future non-local option.

Hmm, what's the current gap with specifying `["security-key", "hybrid"]`? Is it the ergonomics of needing to know to specify both values to achieve equivalent support to `authenticatorAttachment: "cross-platform"`?

**Comment by Firstyear** [comment]
As far as I recall, even in older version authenticatorAttachment has always been a "hint" as well. It just so happens it predictably works today, but it's still a hint. So changing this from platform/cross-platform to more specific transports / factors like "security-key" or "hybrid" is a good change.

**Comment by MasterKale** [comment]
> @MasterKale, order matters with hints; so to be clear, you were intentional in placing `"security-key"` before `"hybrid"`, right?

No, I was not intentionally ordering the hints in that comment. RP's should be free to order hints as they see fit.

> Perhaps it was an oversight, but [`PublicKeyCredential`](https://www.w3.org/TR/webauthn-3/#publickeycredential) only has a field for `authenticatorAttachment`. If `PublicKeyCredentialHints` is designed to be a replacement, then ideally `PublicKeyCredential` would be updated to contain a `hints` `attribute`.

`authenticatorAttachment` still has value as a signal of whether the user might need to be prompted to register their local access device. Hints don't control browser behavior in the same way that authenticatorAttachment does; it's possible that specifying `hints: ["security-key"]` leads to the user completing hybrid authentication anyway. What value might be returned for a hypothetical "`PublicKeyCredential.hints`" in the response in this scenario? `["security-key"]` wouldn't make sense, but the RP didn't specify `["hybrid"]` so it could be confusing...meanwhile returning `authenticatorAttachment: "cross-platform"` remains an accurate description of the general type of the authenticator that ultimately completed the ceremony, so I can't see the benefit of replacing `PublicKeyCredential.authenticatorAttachment` right now.

> According to [@MasterKale's comment](https://github.com/w3c/webauthn/issues/2053#issuecomment-2023398238), the inverse seems to be `"cross-platform"` → `["security-key", "hybrid"]` and `"platform"` → `["client-device"]`; however if that is the case, it should be mentioned in the spec.

Mapping in the opposite direction, from `hints` to `authenticatorAttachment`, isn't quite that easy. While `["hybrid", "security-key"]` and `["security-key", "hybrid"]` conveys slightly different desires by the RP about which type of authenticator they want the browser to start the user in. Naively mapping either combination to `"cross-platform"` prunes this desire. 

---

My ultimate goal with hints was to make it possible to break up registration into three separate flows. The introduction of passkeys jammed hybrid registration into previously-security-key-only flows. For RP's (like Duo Security) that want to be able to show the user instructions before the WebAuthn ceremony this made it incredibly difficult to pull this off. Especially when we wanted the user to use their security key but were shown this by their browser:

![Screenshot 2024-04-19 at 2 23 23 PM](https://github.com/w3c/webauthn/assets/5166470/f572cd92-676b-459c-8e98-07b40be3ef0d)

Hints enable registration to now be three options, as passing in `["security-key"]` can have the browser jump straight to security key interaction when `.create()` is called:

![Screenshot 2024-04-19 at 2 23 33 PM](https://github.com/w3c/webauthn/assets/5166470/0adac5d5-14fe-449f-b949-7e99bbfc68e0)

Sure, clients are free to allow the user to choose a different kind of authenticator than specified in the hint, but if users are just looking to click Next to get through the first thing shown to them this doesn't seem too bad a restriction over using `authenticatorAttachment`.

**Comment by MasterKale** [comment]
> This means that the countably infinite `[PublicKeyCredentialHints]?` is partitioned into five equivalence classes, correct:
> 
> 1. `[]`
> 2. `["security-key"]`
> 3. `["client-device"]`
> 4. `["security-key", "hybrid"]`
> 5. `["client-device", "hybrid"]`

I would suggest that most RP's either A) use hints one-at-a-time to offer pre-registration guidance for the general category of authenticator they want the user to register, or B) group hints along the current `authenticatorAttachment` split:

1. ["client-device"]
2. ["security-key", "hybrid"]

IMO number 2 could just as easily put `"hybrid"` before `"security-key"` as the RP desires - nothing about hints forces any of them into buckets, that should be an opinion established by RP's (or library maintainers.)

> Here I am assuming that a missing `hints` field, `["hybrid"]`, and `[]` are the same.

Not necessarily. `["hybrid"]` is the RP communicating to browsers that support hints, "please start the user in your hybrid registration flow." `[]` carries no meaning and so the browser would be free to start the registration ceremony where it wants to (e.g. the curernt behavior of optimizing for platform authenticator registration.)

**Comment by MasterKale** [comment]
> Is this true? The spec says "if two hints are contradictory, the first one controls" which seems to imply that something like `["security-key", "client-device"]` is equivalent to `["security-device"]` due to the two contradicting each other and `"security-key"` appearing first.

I'd suggest that providing two places is nice because it gives priority on the first, but also provides browsers with direction if the first method isn't available.

For example, a browser could receive `["client-device", "hybrid"]`, know that a platform is unavailable in its own way outside the scope of this discussion, and decide to pop up the hybrid QR code instead because the browser knew it had enough prerequisites satisfied to perform a hybrid ceremony. Without the inclusion of `"hybrid"` the browser might have otherwise shown both hybrid and security keys when the RP isn't as interested in exposing security keys to their users.

I'll have to respond to your other points later, I'm heading back home from today's WAWG F2F ✌️

**Comment by MasterKale** [comment]
**From WAWG meeting @ 5/1:** there's potential value in keeping `authenticatorAttachment` around while making `hints` into a further refinement of existing client behavior for a given attachment. This would mean updating the spec to declare valid hints per attachment (e.g. "for `"cross-platform"` attachment valid hints include `"security-key"` and `"hybrid"`...")

**Comment by emlun** [comment]
> Doesn't `"hybrid"` mean both `"platform"` and `"cross-platform"` are allowed without preference to either one?

No, the [`"hybrid"` hint](https://www.w3.org/TR/webauthn-3/#dom-publickeycredentialhints-hybrid) indicates a preference for a separate authenticator device using the [`"hybrid"` transport](https://www.w3.org/TR/webauthn-3/#dom-authenticatortransport-hybrid), meaning it is separate from the [client device](https://www.w3.org/TR/webauthn-3/#client-device) and therefore [cross-platform attached](https://www.w3.org/TR/webauthn-3/#cross-platform-attachment).

You may be thinking of the use case for `hints: ["client-device", "hybrid"]`, which signals a preference for essentially anything except a dedicated external security key: an integrated platform authenticator or (most likely) the platform authenticator on another nearby general-purpose computing device such as a smartphone (mirroring the [`"passkeyPlatformAuthenticator"` client capability](https://w3c.github.io/webauthn/#dom-clientcapability-passkeyplatformauthenticator)).

> The spec says "if two hints are contradictory, the first one controls" which seems to imply that something like `["security-key", "client-device"]` is equivalent to `["security-device"]` due to the two contradicting each other and `"security-key"` appearing first.

Immediately preceding that, the spec says:

>Hints are provided in order of decreasing preference [...]

so no, your two examples are not equivalent - the former expresses a fallback preference of `"client-device"` while the latter expresses no fallback preference.

For the purposes of mapping to a suitable value of `authenticatorSelection.authenticatorAttachment`, though: yes, in this case the first hint would determine the respective `authenticatorAttachment` setting, and unconditionally override any and all other hints following it. So as of now there are 4 partitions, as you put it: `["security-key", ...]` and `["hybrid", ...]` mapping to `"cross-platform"`, `["client-device", ...]` mapping to `"platform"` and `[]` mapping to `undefined` (absent).

**Comment by MasterKale** [comment]
From WG meeting: It sounds like the best path forward here is to map `authenticatorAttachment` to specific combinations of `hints` values, then allow RPs to continue using `authenticatorAttachment` while encouraging new RPs to use `hints` instead. `hints` would continue to take precedence over `authenticatorAttachment` when both are present. Then we could add a non-normative note to `authenticatorAttachment` suggesting using `hints` instead (essentially a "soft deprecation" without intention of removing `authenticatorAttachment` entirely.)

@dveditz @pascoej: what do you think about this proposed way of encouraging more RPs to use hints over specifying an attachment? Attachments are overloaded and confusing these days especially with synced passkeys and the many third-party providers intercepting API calls or interfacing with the platform. Hints can help RPs better adapt their use of WebAuthn to be flexible in light of all this complexity.

**Comment by emlun** [comment]
Agreed, although I'd map `hybrid` to `cross-platform` instead:

- `platform` : `["internal"]`
- `cross-platform`: `["usb", "nfc", "ble", "hybrid"]`

because the main characteristic of a platform credential is that it integrated into _this_ (current) client device, not just _some_ client device. This is what you want if you're using `attachment` (both the parameter and the output) to suggest registering a platform credential on _this_ device for easier future re-authentication, or to suggest registering a cross-platform credential for access on other client devices.

**Comment by Gunni** [comment]
Why can a website decide if I can use my security key or force me to use the platform tpm thing instead?

Why is that any of their business?

Hint is fine, force is not, imo...
